Docker.DotNet/Microsoft.Net.Http.Client/HttpConnection.cs

```diff
@@ -0,0 +1,169 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Net.Http.Client
+{
+    internal class HttpConnection : IDisposable
+    {
+        private const string CRLF = "\r\n";
+
+        public HttpConnection(BufferedReadStream transport)
+        {
+            Transport = transport;
+        }
+
+        public BufferedReadStream Transport { get; private set; }
+
+        public async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            try
+            {
+                // Serialize headers & send
+                string rawRequest = SerializeRequest(request);
+                byte[] requestBytes = Encoding.ASCII.GetBytes(rawRequest);
+                await Transport.WriteAsync(requestBytes, 0, requestBytes.Length, cancellationToken);
+
+                if (request.Content != null)
+                {
+                    if (request.Content.Headers.ContentLength.HasValue)
+                    {
+                        await request.Content.CopyToAsync(Transport);
+                    }
+                    else
+                    {
+                        // The length of the data is unknown. Send it in chunked mode.
+                        using (var chunkedStream = new ChunkedWriteStream(Transport))
+                        {
+                            await request.Content.CopyToAsync(chunkedStream);
+                            await chunkedStream.EndContentAsync(cancellationToken);
+                        }
+                    }
+                }
+
+                // Receive headers
+                List<string> responseLines = await ReadResponseLinesAsync(cancellationToken);
+                // Determine response type (Chunked, Content-Length, opaque, none...)
+                // Receive body
+                return CreateResponseMessage(responseLines);
+            }
+            catch (Exception ex)
+            {
+                Dispose(); // Any errors at this layer abort the connection.
```

Comment:
Im just not used to this pattern of disposing from within the call so pointing out again. Are we sure this is the right approach?